style: Fix dynamic changes of attributes when combined with :not.

emilio opened the pull request:
This makes the dependency tracker properly recurse into simple selectors inside the current complex selector to find the appropriate dependencies.

We can't still remove the outer visitor because we need it for stuff like `:not(.foo + bar)`, but I plan to get rid of it in a followup as long as try comes back green.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/16364)
<!-- Reviewable:end -->

emilio committed:
style: Fix dynamic changes of attributes when combined with :not.

highfive commented:
Heads up! This PR modifies the following files:
 * @bholley: components/style/restyle_hints.rs

emilio committed:
style: Simplify dependency visitor, avoid tracking dependencies of nested complex selectors separately.

bholley commented on components/style/restyle_hints.rs:
```diff
@@ -421,6 +421,8 @@ fn needs_cache_revalidation(sel: &SimpleSelector<SelectorImpl>) -> bool {
         SimpleSelector::OnlyOfType => true,
+        // FIXME(emilio): This sets the "revalidation" flag for :any, which is
+        // probably expensive given we use it a lot in UA sheets.
         SimpleSelector::NonTSPseudoClass(ref p) => p.state_flag().is_empty(),
```
Please file a perf bug about this.

bholley commented on components/style/restyle_hints.rs:
```diff
@@ -606,6 +554,52 @@ impl DependencySet {
+                ss.visit(&mut sensitivities_visitor);
+            }
+
+            sensitivities_visitor.hint |= combinator_to_restyle_hint(combinator);
```
nit: I think it would be a bit clearer to do |let hint = sensitivities_visitor.hint | combinator_to_restyle_hint(combinator)| to make it clear that the value isn't going to be reused inside sensitivities_visitor.

bholley commented:
@bors-servo delegate+

bors-servo commented:
:v: @emilio can now approve this pull request

emilio committed:
style: Smoke-test the dependency tracking logic.

MozReview-Commit-ID: J5HWdS1H49s
Signed-off-by: Emilio Cobos Álvarez <emilio@crisal.io>

emilio committed:
stylo: Fix restyle hints check.

It's clearly wrong to use intersects() there, since the coming check may be a
bigger hint.

MozReview-Commit-ID: 1YZ8OD0ajSz
Signed-off-by: Emilio Cobos Álvarez <emilio@crisal.io>

emilio commented:
@bors-servo r=bholley

 * Figured out the regressions this patch caused. Was a very embarrassing bug I introduced in 3e80f482ba9 (whoops).

bors-servo commented:
:pushpin: Commit 16cc441 has been approved by `bholley`

<!-- @bors-servo r=bholley 16cc441d44171ba8b72c743c2bc75d0a03b078c2 -->

emilio commented:
@bors-servo r=bholley p=1

 * The damn manifest

bors-servo commented:
:pushpin: Commit e5dff21 has been approved by `bholley`

<!-- @bors-servo r=bholley e5dff21d22203b82896916ccf3387377382ae94d -->

bors-servo commented:
:hourglass: Testing commit e5dff21d22203b82896916ccf3387377382ae94d with merge 1b4e32ec631a97eca341b33ec8f37fd7ba860f93...

bors-servo commented:
:broken_heart: Test failed - [linux-dev](http://build.servo.org/builders/linux-dev/builds/6635)

emilio commented:
@bors-servo r=bholley

 * Don't compile the test on Gecko mode, because of pseudo-element symbols not present.

bors-servo commented:
:pushpin: Commit fcce61e has been approved by `bholley`

<!-- @bors-servo r=bholley fcce61ee104131b25d92c40dd99335533b701c8b -->

bors-servo commented:
:hourglass: Testing commit fcce61ee104131b25d92c40dd99335533b701c8b with merge 2a7b357bfd5bae01e4f5903612d455c0dcbe983c...

bors-servo commented:
:broken_heart: Test failed - [android](http://build.servo.org/builders/android/builds/6294)

bholley commented:
@bors-servo retry

infra

bors-servo commented:
:hourglass: Testing commit fcce61ee104131b25d92c40dd99335533b701c8b with merge 53c47acfc425ac2dcc5d8aa08d4882c4bb0e7251...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/6297), [arm32](http://build.servo.org/builders/arm32/builds/6509), [arm64](http://build.servo.org/builders/arm64/builds/6544), [linux-dev](http://build.servo.org/builders/linux-dev/builds/6645), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/3382), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/3405), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/6649), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/6686), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/3335), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/3406), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/1822)
Approved by: bholley
Pushing 53c47acfc425ac2dcc5d8aa08d4882c4bb0e7251 to master...
